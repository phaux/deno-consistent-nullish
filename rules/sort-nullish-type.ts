import { fc } from "../util/formatCode.ts";
import { getRangeWithParens } from "../util/getRangeWithParens.ts";

const typeOrder: Partial<Record<Deno.lint.TypeNode["type"], number>> = {
  TSNullKeyword: 1,
  TSUndefinedKeyword: 2,
};

export const sortNullishType: Deno.lint.Rule = {
  create: (ctx) => ({
    TSUnionType(unionNode) {
      const sortedTypes = unionNode.types
        .filter((t) => (typeOrder[t.type] ?? 0) > 0)
        .sort((a, b) => typeOrder[a.type]! - typeOrder[b.type]!)
        .map((t) => ctx.sourceCode.getText(t))
        .join(" | ");

      for (const [index, node] of unionNode.types.entries()) {
        if ((typeOrder[node.type] ?? 0) <= 0) continue;

        const targetIndex = unionNode.types.findLastIndex(
          (t) => (typeOrder[t.type] ?? 0) < (typeOrder[node.type] ?? 0),
        );
        if (targetIndex < index) continue;

        const targetNode = unionNode.types[targetIndex]!;
        const nodeText = ctx.sourceCode.getText(node);
        const targetNodeText = ctx.sourceCode.getText(targetNode);

        ctx.report({
          node: node,
          message: `Expected ${fc(nodeText)} to be after ${fc(targetNodeText)}`,
          hint: `For consistency, place ${fc(sortedTypes)} at the end of the union.`,
          *fix(fixer) {
            // Start from the end of previous member or start of the union if first member.
            const start = index === 0 ? unionNode.range[0] : getRangeWithParens(
              ctx.sourceCode,
              unionNode.types[index - 1]!,
            )[1];
            // End at the end of the current member.
            const end = getRangeWithParens(
              ctx.sourceCode,
              unionNode.types[index]!,
            )[1];
            // Remove range.
            yield fixer.removeRange([start, end]);
            // Insert current node text after the target node.
            yield fixer.insertTextAfterRange(
              getRangeWithParens(ctx.sourceCode, targetNode),
              ` | ${nodeText}`,
            );
          },
        });
      }
    },
  }),
};
