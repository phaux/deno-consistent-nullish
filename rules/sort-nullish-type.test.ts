import { assertSnapshot } from "@std/testing/snapshot";
import { assertEquals } from "@std/assert";
import { sortNullishType } from "./sort-nullish-type.ts";

const plugin: Deno.lint.Plugin = {
  name: "test",
  rules: { rule: sortNullishType },
};

Deno.test("bans wrong order of types", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    interface A { a: null | string; }
    type B = { b?: undefined | number; }
    function foo(arg?: boolean | undefined | null): undefined | null {}
    let x: T | null | void | undefined;
    const y: T | undefined | null | void;
    `,
  );
  assertEquals(diagnostics.length, 7);
  await assertSnapshot(t, diagnostics);
});

Deno.test("allows correct order of types", () => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    interface A { a: string | null; }
    type B = { b?: number | undefined; }
    function foo(arg?: boolean | null | undefined): null | undefined {}
    let x: T | void | null | undefined;
    const y: T | void | null | undefined;
    `,
  );
  assertEquals(diagnostics.length, 0);
});
