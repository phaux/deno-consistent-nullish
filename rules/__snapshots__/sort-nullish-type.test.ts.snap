export const snapshot = {};

snapshot[`bans wrong order of types 1`] = `
[
  {
    fix: [
      {
        range: [
          22,
          26,
        ],
        text: "",
      },
      {
        range: [
          35,
          35,
        ],
        text: " | null",
      },
    ],
    hint: "For consistency, place \`null\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`null\` to be after \`string\`",
    range: [
      22,
      26,
    ],
  },
  {
    fix: [
      {
        range: [
          58,
          67,
        ],
        text: "",
      },
      {
        range: [
          76,
          76,
        ],
        text: " | undefined",
      },
    ],
    hint: "For consistency, place \`undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`undefined\` to be after \`number\`",
    range: [
      58,
      67,
    ],
  },
  {
    fix: [
      {
        range: [
          132,
          141,
        ],
        text: "",
      },
      {
        range: [
          148,
          148,
        ],
        text: " | undefined",
      },
    ],
    hint: "For consistency, place \`null | undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`undefined\` to be after \`null\`",
    range: [
      132,
      141,
    ],
  },
  {
    fix: [
      {
        range: [
          110,
          122,
        ],
        text: "",
      },
      {
        range: [
          129,
          129,
        ],
        text: " | undefined",
      },
    ],
    hint: "For consistency, place \`null | undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`undefined\` to be after \`null\`",
    range: [
      113,
      122,
    ],
  },
  {
    fix: [
      {
        range: [
          164,
          171,
        ],
        text: "",
      },
      {
        range: [
          178,
          178,
        ],
        text: " | null",
      },
    ],
    hint: "For consistency, place \`null | undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`null\` to be after \`void\`",
    range: [
      167,
      171,
    ],
  },
  {
    fix: [
      {
        range: [
          206,
          218,
        ],
        text: "",
      },
      {
        range: [
          232,
          232,
        ],
        text: " | undefined",
      },
    ],
    hint: "For consistency, place \`null | undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`undefined\` to be after \`void\`",
    range: [
      209,
      218,
    ],
  },
  {
    fix: [
      {
        range: [
          218,
          225,
        ],
        text: "",
      },
      {
        range: [
          232,
          232,
        ],
        text: " | null",
      },
    ],
    hint: "For consistency, place \`null | undefined\` at the end of the union.",
    id: "test/rule",
    message: "Expected \`null\` to be after \`void\`",
    range: [
      221,
      225,
    ],
  },
]
`;
